ftw/mail/upgrades/20210702104300_include_missing_em_ls_in_attachments/upgrade.py

```diff
@@ -0,0 +1,36 @@
+from ftw.mail import utils
+from ftw.upgrade import UpgradeStep
+import logging
+
+LOG = logging.getLogger('ftw.upgrade')
+
+
+class IncludeMissingEMLsInAttachments(UpgradeStep):
+    """Include missing emls in attachments.
+    """
+
+    deferrable = True
+
+    def __call__(self):
+        query = {'portal_type': 'ftw.mail.mail'}
+        for mail in self.objects(query, 'Include missing emls in attachments'):
+            self.add_missing_eml_attachments(mail)
+
+    def add_missing_eml_attachments(self, obj):
+        attachments = utils.get_attachments(obj.msg)
+        if len(attachments) == len(obj.attachment_infos):
+            return
+
+        info_list = list(obj.attachment_infos)
+        stored_positions = [each['position'] for each in info_list]
+
+        for attachment in attachments:
+            if attachment['position'] in stored_positions:
+                continue
+            if (attachment['filename'] != "attachment.eml" or
+                    attachment['content-type'] != "message/rfc822"):
+                LOG.warning("Found missing attachment that is not in-line eml"
```

Comment:
this uses format with a non unicode format string. it may fail if one of the format parameters (`attachment`) are non ascii and contain special characters.